module/os/linux/zfs/zpl_file.c

```diff
@@ -161,21 +173,40 @@ static int
 zpl_fsync(struct file *filp, loff_t start, loff_t end, int datasync)
 {
 	struct inode *inode = filp->f_mapping->host;
+	znode_t *zp = ITOZ(inode);
+	zfsvfs_t *zfsvfs = ITOZSB(inode);
 	cred_t *cr = CRED();
 	int error;
 	fstrans_cookie_t cookie;
 
+	ZPL_ENTER(zfsvfs);
+	ZPL_VERIFY_ZP(zp);
+	atomic_inc_32(&zp->z_sync_writes_cnt);
+	/*
+	 * speed up any non-sync page writebacks since
+	 * they may take several seconds to complete
+	 */
+	if (atomic_load_32(&zp->z_async_writes_cnt) > 0) {
+		zil_commit(zfsvfs->z_log, zp->z_id);
```

Comment:
Isn't this already handled by the `zil_commmit()` in the  `filemap_write_and_wait_range()->...>do_writepages()->zpl_writepages()` call below?  It not, it still seems like a more appropriate place for this check.

Reply:
@behlendorf originally i did not have this check in the fix but after some testing with the failing test, the issue was still persisting (albeit more rarely).

i finally found out that it was due to this early return here (described as case 2 in the PR description):
https://github.com/torvalds/linux/blob/cb690f5238d71f543f4ce874aa59237cf53a877c/mm/filemap.c#L403

due to this early return, we don't get back into zfs code and still call filemap_write_and_wait_range() which can block for several seconds.

i wasn't too happy about scattering the logic around but i haven't found a simpler way to make the fix work

Reply:
the fix can roughly be reasoned about this way:
- if the sync writeback doesn't "see" that there is an overlapping sync & non-sync writeback, then it is guaranteed that the non-sync writeback will "see" it

Reply:
it's important to increment `z_sync_writes_cnt` before doing the check on `z_async_writes_cnt` to provide this guarantee

Reply:
I wish I saw a better way to handle this myself, but I can't say I do.  Let's at least expand the comment here to make it more clear why this is important and under what circumstances this is needed.

Reply:
fixed: i've added some comments, feel free to let me know if it needs to be trimmed down if it's too much